### PR TITLE
Initiate update on machine template ref change

### LINKED ISSUE #808: Control plane contoller doesn't react on machineTemplate ref changes
### What happened?

If the machineTemplate reference changes, k0smotron doesn't do anything. 

### Involved Areas

_No response_

### What did you expect to happen?

k0smotron should create new machines and remove old ones

### Steps to reproduce

- Deploy a K0sControlPlane cluster
- Change the machineTemplate reference

### k0smotron version

any

### k0s version

any

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by makhov**
Fixes #808

### CODE REVIEW

**Review by apedriza — COMMENTED** [review]
code looks good but I think we have some issue in our machines sync algo. Following the new e2e test added in the PR, **if a second update is applied**, cluster update gets stuck in this state:
Screenshot 2024-11-18 at 11 56 17

I applied the following update:
 ```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: DockerMachineTemplate
metadata:
  name: docker-test-cp-template-new2
  namespace: default
spec:
  template:
    spec:
      customImage: kindest/node:v1.31.1
---
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: K0sControlPlane
metadata:
  name: docker-test
spec:
  replicas: 3
  version: v1.31.2+k0s.0
  updateStrategy: Recreate
  k0sConfigSpec:
    k0s:
      apiVersion: k0s.k0sproject.io/v1beta1
      kind: ClusterConfig
      metadata:
        name: k0s
      spec:
        api:
          extraArgs:
            anonymous-auth: "true"
        telemetry:
          enabled: false
  machineTemplate:
    infrastructureRef:
      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
      kind: DockerMachineTemplate
      name: docker-test-cp-template-new2
      namespace: default
```

**Comment by makhov** [comment]
@apedriza yeah, thanks! I fixed the issues and added second update to the test case.

File: internal/controller/controlplane/k0s_controlplane_controller.go [hunk 1]
```diff
@@ -242,64 +249,67 @@ func (c *K0sController) reconcileKubeconfig(ctx context.Context, cluster *cluste
...
+		} else {
+			desiredMachineNames[m.Name] = true
 		}
 	}
+	log.Log.Info("Collected machines", "count", machines.Len(), "desired", kcp.Spec.Replicas, "updating", clusterIsUpdating, "deleting", len(machineNamesToDelete), "desiredMachines", desiredMachineNames)
 
-	if oldMachines > 0 {
+	if clusterIsUpdating {
 		log.Log.Info("Cluster is updating", "currentVersion", currentVersion, "newVersion", kcp.Spec.Version, "strategy", kcp.Spec.UpdateStrategy)
 		clusterIsUpdating = true
```
[the remark this answers is not in the document]
Comment (makhov) [reply internal/controller/controlplane/k0s_controlplane_controller.go]: Fixed, thanks

File: internal/controller/controlplane/k0s_controlplane_controller.go [hunk 2]
```diff
@@ -375,75 +379,88 @@ func (c *K0sController) reconcileMachines(ctx context.Context, cluster *clusterv
...
+			return fmt.Errorf("error getting cluster client set for deletion: %w", err)
 		}
 
 		// Remove the last machine and report the new number of replicas to status
 		// On the next reconcile, the next machine will be removed
 		// Wait for the previous machine to be deleted to avoid etcd issues
-		machine := machines.Oldest()
+		machine := machines.Filter(func(m *clusterv1.Machine) bool {
+			return machineNamesToDelete[m.Name]
+		}).Newest()
```
[the remark this answers is not in the document]
Comment (makhov) [reply internal/controller/controlplane/k0s_controlplane_controller.go]: It really doesn't matter. I change it back to oldest to keep it consistent

File: internal/controller/controlplane/helper.go [hunk 3]
```diff
@@ -211,15 +247,70 @@ func (c *K0sController) generateMachineFromTemplate(ctx context.Context, name st
...
+	return false, nil
+}
+
 func (c *K0sController) markChildControlNodeToLeave(ctx context.Context, name string, clientset *kubernetes.Clientset) error {
 	if clientset == nil {
 		return nil
 	}
+
 	logger := log.FromContext(ctx).WithValues("controlNode", name)
+	//err := clientset.RESTClient().
```
[the remark this answers is not in the document]
Comment (makhov) [reply internal/controller/controlplane/helper.go]: Fixed, thanks

**Review by apedriza — APPROVED** [review]

**Pull request merged**